File: android/wallet-app/src/main/kotlin/io/github/vsima/canton/app/WalletConnect.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton.app

import android.app.Application
import android.util.Log
import com.reown.android.Core
import com.reown.android.CoreClient
import com.reown.walletkit.client.Wallet
import com.reown.walletkit.client.WalletKit
import io.github.vsima.canton.dapp.DappWallet
import io.github.vsima.canton.dapp.wallet.DappPeer
import io.github.vsima.canton.dapp.wc.Caip
import io.github.vsima.canton.dapp.wc.CantonWalletConnect
import io.github.vsima.canton.dapp.wc.WcRequest
import io.github.vsima.canton.dapp.wc.WcResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json

/**
 * The Reown WalletKit binding: the relay/pairing/session client that carries
 * CIP-0103 frames to and from the SDK's [CantonWalletConnect] adapter.
 *
 * This is the one place that depends on Reown. It owns no protocol logic —
 * `onSessionProposal` approves with the adapter's namespaces, `onSessionRequest`
 * routes into the adapter's `handle`, and the adapter drives the engine
 * (`DappSession`) which does the approvals and signing. The wallet's identity
 * and the approval UI stay in [WalletModel]; nothing here touches a key.
 *
 * Reown delivers its callbacks on a background thread, so requests are handled
 * on an IO scope; the engine's approval delegate hops to Main to raise the
 * sheet.
 */
/** A live WalletConnect session, for display on the Connect screen. */
data class WcSessionInfo(val topic: String, val name: String, val url: String)

object WalletConnectController {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var networkId: String? = null
    private var adapterFactory: ((DappPeer) -> CantonWalletConnect)? = null
    private var accounts: (suspend () -> List<DappWallet>)? = null

    /** One adapter (and so one `DappSession` + grant) per session topic. */
    private val adapters = mutableMapOf<String, CantonWalletConnect>()

    /** Peer display names by topic, for the disconnect status line: the
     *  session is already gone from WalletKit when its delete event fires. */
    private val peerNames = mutableMapOf<String, String>()

    /** Set by [WalletModel] to surface status lines on the Connect screen. */
    var onStatus: ((String) -> Unit)? = null

    /** Set by [WalletModel] to surface the active sessions on the Connect screen. */
    var onSessions: ((List<WcSessionInfo>) -> Unit)? = null

    /**
     * Registers the accounts the wallet may share and the per-peer adapter
     * factory. The factory runs on a peer's first request, with the peer
     * built from that WalletConnect session's own metadata; see [adapterFor].
     */
    fun register(
        networkId: String,
        accounts: suspend () -> List<DappWallet>,
        adapterFactory: (DappPeer) -> CantonWalletConnect,
    ) {
        this.networkId = networkId
        this.accounts = accounts
        this.adapterFactory = adapterFactory
        synchronized(adapters) { adapters.clear() }
    }

    /**
     * The adapter for [topic], created on first use with the peer identity
     * the transport can attest: the WalletConnect session's peer metadata
     * (self-reported, so `verified` only when Reown's Verify API vouched for
     * the origin in [verify]).
     */
    private fun adapterFor(topic: String, verify: Wallet.Model.VerifyContext?): CantonWalletConnect? {
        val factory = adapterFactory ?: return null
        synchronized(adapters) {
            return adapters.getOrPut(topic) {
                val meta = try {
                    WalletKit.getActiveSessionByTopic(topic)?.metaData
                } catch (e: Throwable) {
                    Log.i("WALLET", "WC: session lookup for $topic failed: $e")
                    null
                }
                val name = meta?.name?.takeIf { it.isNotBlank() } ?: "Unidentified dApp"
                peerNames[topic] = name
                factory(
                    DappPeer(
                        id = topic,
                        name = name,
                        url = meta?.url?.takeIf { it.isNotBlank() },
                        iconUrl = meta?.icons?.firstOrNull { it.isNotBlank() },
                        verified = verify?.validation == Wallet.Model.Validation.VALID,
                    ),
                )
            }
        }
    }

    /** Reads WalletKit's active sessions and pushes them to the UI. */
    fun refreshSessions() {
        val sessions = try {
            WalletKit.getListOfActiveSessions().map { s ->
                WcSessionInfo(topic = s.topic, name = s.metaData?.name ?: "dApp", url = s.metaData?.url ?: "")
            }
        } catch (e: Throwable) {
            Log.i("WALLET", "WC: getListOfActiveSessions failed: $e")
            emptyList()
        }
        onSessions?.invoke(sessions)
    }

    /** Disconnects a session by topic. */
    fun disconnect(topic: String) {
        synchronized(adapters) { adapters.remove(topic) }
        WalletKit.disconnectSession(
            Wallet.Params.SessionDisconnect(sessionTopic = topic),
            onSuccess = { refreshSessions() },
            onError = { error -> status("Disconnect failed: ${error.throwable.message}") },
        )
    }

    /** The dApp side (or the relay) ended a session. */
    fun onSessionDelete(delete: Wallet.Model.SessionDelete) {
        if (delete is Wallet.Model.SessionDelete.Success) {
            val name = synchronized(adapters) {
                adapters.remove(delete.topic)
                peerNames.remove(delete.topic)
            }
            status("${name ?: "A dApp"} disconnected")
        }
        refreshSessions()
    }

    /** Hands a `wc:` pairing URI to the relay. */
    fun pair(uri: String) {
        WalletKit.pair(Wallet.Params.Pair(uri)) { error ->
            status("Pairing failed: ${error.throwable.message}")
        }
    }

    fun onSessionProposal(proposal: Wallet.Model.SessionProposal) {
        val networkId = networkId
        val accounts = accounts
        if (networkId == null || accounts == null || adapterFactory == null) {
            reject(proposal, "Wallet not ready")
            return
        }
        scope.launch {
            try {
                // Approve the methods the dApp asked for that the engine can
                // serve: the ecosystem proposes canton_-prefixed names, a
                // CIP-0103-verbatim dApp proposes bare ones, and both clients
                // refuse any request outside the approved set.
                val requested = (proposal.requiredNamespaces.values + proposal.optionalNamespaces.values)
                    .flatMap { it.methods }
                val ns = CantonWalletConnect.sessionNamespaces(Caip.chainId(networkId), accounts(), requested)
                val namespaces = mapOf(
                    Caip.CANTON_NAMESPACE to Wallet.Model.Namespace.Session(
                        chains = ns.chains,
                        methods = ns.methods,
                        events = ns.events,
                        accounts = ns.accounts,
                    ),
                )
                WalletKit.approveSession(
                    Wallet.Params.SessionApprove(
                        proposerPublicKey = proposal.proposerPublicKey,
                        namespaces = namespaces,
                    ),
                    onSuccess = {
                        status("Connected to ${proposal.name.ifBlank { "dApp" }}")
                        refreshSessions()
                    },
                    onError = { error -> status("Approve failed: ${error.throwable.message}") },
                )
            } catch (e: Exception) {
                status("Proposal error: ${e.message}")
            }
        }
    }

    fun onSessionRequest(request: Wallet.Model.SessionRequest, verify: Wallet.Model.VerifyContext?) {
        val adapter = adapterFor(request.topic, verify) ?: return
        scope.launch {
            val id = request.request.id
            val topic = request.topic
            val params = runCatching { Json.parseToJsonElement(request.request.params) }.getOrNull()
            val response = adapter.handle(
                WcRequest(
                    topic = topic,
                    requestId = id,
                    chainId = request.chainId ?: "",
                    method = request.request.method,
                    params = params,
                ),
            )
            val jsonRpc = when (response) {
                is WcResponse.Success ->
                    Wallet.Model.JsonRpcResponse.JsonRpcResult(id = id, result = response.result.toString())
                is WcResponse.Error ->
                    Wallet.Model.JsonRpcResponse.JsonRpcError(id = id, code = response.code, message = response.message)
            }
            WalletKit.respondSessionRequest(
                Wallet.Params.SessionRequestResponse(sessionTopic = topic, jsonRpcResponse = jsonRpc),
                onSuccess = {},
                onError = { error -> status("Respond failed: ${error.throwable.message}") },
            )
        }
    }

    private fun reject(proposal: Wallet.Model.SessionProposal, reason: String) {
        WalletKit.rejectSession(
            Wallet.Params.SessionReject(proposerPublicKey = proposal.proposerPublicKey, reason = reason),
            onSuccess = {},
            onError = {},
        )
    }

    private fun status(line: String) {
        Log.i("WALLET", "WC: $line")
        onStatus?.invoke(line)
    }
}

/**
 * The wallet's [Application], which initialises Reown WalletKit once for the
 * process and forwards its delegate callbacks to [WalletConnectController]. Init
 * must happen here, before any Activity, and Core before WalletKit.
 */
class WalletApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        val metadata = Core.Model.AppMetaData(
            name = "Canton Wallet",
            description = "Canton reference wallet",
            url = "https://github.com/vsima/canton-mobile-app",
            icons = emptyList(),
            redirect = "canton-wallet://wc",
        )
        CoreClient.initialize(
            application = this,
            projectId = WC_PROJECT_ID,
            metaData = metadata,
            onError = { error -> Log.i("WALLET", "WC core init failed: ${error.throwable}") },
        )
        WalletKit.initialize(
            Wallet.Params.Init(core = CoreClient),
            onSuccess = { Log.i("WALLET", "WC WalletKit ready") },
            onError = { error -> Log.i("WALLET", "WC WalletKit init failed: ${error.throwable}") },
        )
        WalletKit.setWalletDelegate(WalletDelegate)
    }

    private object WalletDelegate : WalletKit.WalletDelegate {
        override fun onSessionProposal(
            sessionProposal: Wallet.Model.SessionProposal,
            verifyContext: Wallet.Model.VerifyContext,
        ) = WalletConnectController.onSessionProposal(sessionProposal)

        override fun onSessionRequest(
            sessionRequest: Wallet.Model.SessionRequest,
            verifyContext: Wallet.Model.VerifyContext,
        ) = WalletConnectController.onSessionRequest(sessionRequest, verifyContext)

        override val onSessionAuthenticate:
            ((Wallet.Model.SessionAuthenticate, Wallet.Model.VerifyContext) -> Unit)? = null

        override fun onSessionDelete(sessionDelete: Wallet.Model.SessionDelete) {
            WalletConnectController.onSessionDelete(sessionDelete)
        }
        override fun onSessionExtend(session: Wallet.Model.Session) {}
        override fun onSessionSettleResponse(response: Wallet.Model.SettledSessionResponse) {}
        override fun onSessionUpdateResponse(response: Wallet.Model.SessionUpdateResponse) {}
        override fun onProposalExpired(proposal: Wallet.Model.ExpiredProposal) {}
        override fun onRequestExpired(request: Wallet.Model.ExpiredRequest) {}
        override fun onConnectionStateChange(state: Wallet.Model.ConnectionState) {}
        override fun onError(error: Wallet.Model.Error) {}
    }

    companion object {
        /** Non-secret WalletConnect project id (a public client key). */
        const val WC_PROJECT_ID = "cbef3d23404e895fdc178fadcf6798c1"
    }
}

File: ios/Sources/Wallet/WalletConnect.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

import Combine
import Foundation
import CantonDappKit
import CantonDappWCKit
// RPCID / RPCResult / JSONRPCError are not Sendable-audited either; treat them
// as pre-concurrency so an `RPCID` can be handed to a nonisolated responder.
@preconcurrency import JSONRPC
// Reown is not Sendable-audited for Swift 6, so its singleton `WalletKit.instance`
// trips complete-concurrency checking. @preconcurrency relaxes those diagnostics
// for Reown symbols only — the rest of the app stays under Swift 6. Reown's
// client is internally synchronized and only touched here on the main actor.
@preconcurrency import ReownWalletKit
import WalletConnectNetworking

/// A live WalletConnect session, for display on the Connect screen.
struct WcSessionInfo: Identifiable, Equatable {
    var id: String { topic }
    let topic: String
    let name: String
    let url: String
}

/// The Reown WalletKit binding: the relay/pairing/session client that carries
/// CIP-0103 frames to and from the SDK's ``CantonWalletConnect`` adapter.
///
/// This is the one place that depends on Reown. It owns no protocol logic —
/// a session proposal is approved with the adapter's namespaces, a session
/// request is routed into the adapter's `handle`, and the adapter drives the
/// engine (`DappSession`) which does the approvals and signing. The wallet's
/// identity and the approval UI stay in ``WalletModel``; nothing here touches a
/// key. It is the iOS twin of Android's `WalletConnectController`.
@MainActor
final class WalletConnectController {
    static let shared = WalletConnectController()
    private init() {}

    private var adapter: CantonWalletConnect?
    private var accountsProvider: (@Sendable () async -> [DappWallet])?
    private var cancellables = Set<AnyCancellable>()
    private var configured = false

    /// Set by ``WalletModel`` to surface status lines on the Connect screen.
    var onStatus: ((String) -> Void)?
    /// Set by ``WalletModel`` to surface the active sessions on the Connect screen.
    var onSessions: (([WcSessionInfo]) -> Void)?

    /// Initialises Reown once for the process and subscribes to its publishers.
    /// Must run before any pairing; safe to call repeatedly.
    func configure(projectId: String) {
        guard !configured else { return }
        configured = true
        do {
            let metadata = AppMetadata(
                name: "Canton Wallet",
                description: "Canton reference wallet",
                url: "https://github.com/vsima/canton-mobile-app",
                icons: [],
                redirect: try AppMetadata.Redirect(native: "canton-wallet://", universal: nil)
            )
            Networking.configure(
                groupIdentifier: "group.io.github.vsima.canton.app",
                projectId: projectId,
                socketFactory: URLSessionWebSocketFactory()
            )
            WalletKit.configure(metadata: metadata, crypto: CantonCryptoProvider())
            subscribe()
        } catch {
            print("WALLET: WC configure failed: \(error)")
        }
    }

    private func subscribe() {
        WalletKit.instance.sessionProposalPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] output in self?.handleProposal(output.proposal) }
            .store(in: &cancellables)
        WalletKit.instance.sessionRequestPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] output in self?.handleRequest(output.request) }
            .store(in: &cancellables)
        WalletKit.instance.sessionsPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] sessions in self?.publish(sessions) }
            .store(in: &cancellables)
    }

    /// Registers the wallet's adapter + the accounts it may share.
    func register(adapter: CantonWalletConnect, accounts: @escaping @Sendable () async -> [DappWallet]) {
        self.adapter = adapter
        self.accountsProvider = accounts
        refreshSessions()
    }

    /// Hands a scanned/pasted `wc:` pairing URI to the relay.
    func pair(_ uri: String) {
        let trimmed = uri.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let wcURI = try? WalletConnectURI(uriString: trimmed) else {
            onStatus?("That is not a WalletConnect link")
            return
        }
        Task {
            do { try await WalletKit.instance.pair(uri: wcURI) }
            catch { onStatus?("Pairing failed: \(error.localizedDescription)") }
        }
    }

    /// Reads WalletKit's active sessions and pushes them to the UI.
    func refreshSessions() { publish(WalletKit.instance.getSessions()) }

    /// Disconnects a session by topic.
    func disconnect(topic: String) {
        Task {
            do {
                try await WalletKit.instance.disconnect(topic: topic)
                refreshSessions()
            } catch { onStatus?("Disconnect failed: \(error.localizedDescription)") }
        }
    }

    // MARK: - Reown callbacks

    private func handleProposal(_ proposal: Session.Proposal) {
        guard let adapter, let accountsProvider else {
            reject(proposal, reason: "Wallet not ready")
            return
        }
        let proposalId = proposal.id
        let name = proposal.proposer.name.isEmpty ? "dApp" : proposal.proposer.name
        Task {
            let ns = adapter.sessionNamespaces(accounts: await accountsProvider())
            await approveSession(proposalId: proposalId, name: name, namespaces: ns)
        }
    }

    /// Builds Reown's typed namespaces and approves. `nonisolated` so the
    /// non-Sendable `SessionNamespace` is created and consumed in one isolation
    /// domain, never sent across the actor boundary. Its input
    /// (`WcSessionNamespaces`) and outputs are Sendable.
    private nonisolated func approveSession(proposalId: String, name: String, namespaces ns: WcSessionNamespaces) async {
        guard let namespaces = Self.sessionNamespaces(ns) else {
            await setStatus("Could not build session namespaces")
            return
        }
        do {
            _ = try await WalletKit.instance.approve(proposalId: proposalId, namespaces: namespaces)
            await onApproved(name)
        } catch {
            await setStatus("Approve failed: \(error.localizedDescription)")
        }
    }

    private func handleRequest(_ request: Request) {
        guard let adapter else { return }
        let topic = request.topic
        let requestId = request.id
        let wc = WcRequest(
            topic: request.topic,
            requestId: Self.requestId(request.id),
            chainId: request.chainId.absoluteString,
            method: request.method,
            params: try? Self.jsonValue(from: request.params)
        )
        Task {
            let response = await adapter.handle(wc)
            await sendResponse(topic: topic, requestId: requestId, response: response)
        }
    }

    /// Encodes the CIP-0103 result into Reown's `RPCResult` and responds.
    /// `nonisolated` so the non-Sendable `RPCResult` never crosses isolation.
    private nonisolated func sendResponse(topic: String, requestId: RPCID, response: WcResponse) async {
        do {
            let result: RPCResult
            switch response {
            case .success(let value):
                result = .response(try Self.anyCodable(from: value))
            case .error(let code, let message):
                result = .error(JSONRPCError(code: code, message: message))
            }
            try await WalletKit.instance.respond(topic: topic, requestId: requestId, response: result)
        } catch {
            await setStatus("Respond failed: \(error.localizedDescription)")
        }
    }

    private func onApproved(_ name: String) {
        onStatus?("Connected to \(name)")
        refreshSessions()
    }

    private func setStatus(_ line: String) { onStatus?(line) }

    private func reject(_ proposal: Session.Proposal, reason: String) {
        Task {
            try? await WalletKit.instance.rejectSession(proposalId: proposal.id, reason: .userRejected)
            onStatus?(reason)
        }
    }

    private func publish(_ sessions: [Session]) {
        onSessions?(sessions.map {
            WcSessionInfo(topic: $0.topic, name: $0.peer.name.isEmpty ? "dApp" : $0.peer.name, url: $0.peer.url)
        })
    }

    // MARK: - Mapping helpers

    /// Projects the adapter's ``WcSessionNamespaces`` into Reown's typed CAIP
    /// namespaces. Our CAIP-10 addresses are percent-encoded party ids, which
    /// Reown's `Account` accepts (its address charset includes `%`).
    private nonisolated static func sessionNamespaces(_ ns: WcSessionNamespaces) -> [String: SessionNamespace]? {
        let accounts = ns.accounts.compactMap { Account($0) }
        guard !accounts.isEmpty else { return nil }
        let namespace = SessionNamespace(
            chains: ns.chains.compactMap { Blockchain($0) },
            accounts: accounts,
            methods: Set(ns.methods),
            events: Set(ns.events)
        )
        return [Caip.cantonNamespace: namespace]
    }

    private nonisolated static func requestId(_ id: RPCID) -> Int64 {
        switch id {
        case .right(let value): return value
        case .left(let text): return Int64(text) ?? 0
        }
    }

    /// Reown's `AnyCodable` → the SDK's `JSONValue`, via a JSON round-trip so
    /// neither type's internals leak into the other.
    private nonisolated static func jsonValue(from anyCodable: AnyCodable) throws -> JSONValue {
        try JSONValue.parse(try JSONEncoder().encode(anyCodable))
    }

    private nonisolated static func anyCodable(from value: JSONValue) throws -> AnyCodable {
        try JSONDecoder().decode(AnyCodable.self, from: try value.serialized())
    }
}

/// A `WebSocketConnecting` over `URLSessionWebSocketTask`, so the wallet needs no
/// third-party WebSocket (Starscream) — the relay only needs text frames.
final class URLSessionWebSocket: NSObject, WebSocketConnecting, URLSessionWebSocketDelegate, @unchecked Sendable {
    var request: URLRequest
    var isConnected: Bool = false
    var onConnect: (() -> Void)?
    var onDisconnect: ((Error?) -> Void)?
    var onText: ((String) -> Void)?

    private var task: URLSessionWebSocketTask?
    private lazy var session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)

    init(request: URLRequest) {
        self.request = request
        super.init()
    }

    func connect() {
        let task = session.webSocketTask(with: request)
        self.task = task
        task.resume()
        receive()
    }

    func disconnect() {
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        if isConnected {
            isConnected = false
            onDisconnect?(nil)
        }
    }

    func write(string: String, completion: (() -> Void)?) {
        task?.send(.string(string)) { _ in completion?() }
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                switch message {
                case .string(let text): self.onText?(text)
                case .data(let data): String(data: data, encoding: .utf8).map { self.onText?($0) }
                @unknown default: break
                }
                self.receive()
            case .failure(let error):
                self.isConnected = false
                self.onDisconnect?(error)
            }
        }
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        isConnected = true
        onConnect?()
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        isConnected = false
        onDisconnect?(nil)
    }
}

struct URLSessionWebSocketFactory: WebSocketFactory {
    func create(with url: URL) -> WebSocketConnecting {
        URLSessionWebSocket(request: URLRequest(url: url))
    }
}

/// Reown requires a `CryptoProvider`, but its only use is EVM message recovery
/// (EIP-191 / SIWE `onSessionAuthenticate`), which Canton never uses — the
/// wallet only answers `signMessage` and `prepareExecute`. So this is a stub;
/// implementing it would pull in an EVM crypto stack for no reachable code path.
struct CantonCryptoProvider: CryptoProvider {
    struct Unsupported: Error {}

    func recoverPubKey(signature: EthereumSignature, message: Data) throws -> Data {
        throw Unsupported()
    }

    func keccak256(_ data: Data) -> Data { Data() }
}
